src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java

```diff
@@ -143,16 +178,53 @@ public Builder(String name, String spaceType, String m, String efConstruction) {
 
         @Override
         public KNNVectorFieldMapper build(BuilderContext context) {
-            if (this.spaceType == null) {
-                this.spaceType = getSpaceType(context.indexSettings());
-            }
+            if (knnMethodContext.getValue().getEngine() == KNNEngine.NMSLIB) {
+
+                // Originally, a user would use index settings to set the spaceType, efConstruction and m hnsw
+                // parameters. Upon further review, it makes sense to set these parameters in the mapping of a
+                // particular field. However, because users migrating from older versions will still use the index
+                // settings to set these parameters, we will need to provide backwards compatibilty. In order to
+                // handle this, we first check if the setting is set, and, if so use it. If not, we fall back to
+                // the parameters set in the mapping. This means that if a user sets the settings, setting the mapping
```

Comment:
1. Should this be other way round? First priority is mapping parameters and then index settings. As index settings are deprecated, expectation is to use Mappings. So we would check mappings first and if mapping is not defined(expecting this cluster migrated from old version), we fall back and check index settings?

2. These index settings have default values right, so will they be ```null``` when customer does not set it or simply return default? If it returns default then we might never reach mappings. You may wan to confirm?

Reply:
> 1 Should this be other way round? First priority is mapping parameters and then index settings. As index settings are deprecated, expectation is to use Mappings. So we would check mappings first and if mapping is not defined(expecting this cluster migrated from old version), we fall back and check index settings?

I was thinking about this. We could do that. It would require the default for knnMethodContext to be null. I think this is okay. I will update.

> 2 These index settings have default values right, so will they be null when customer does not set it or simply return default? If it returns default then we might never reach mappings. You may want to confirm?

Confirmed that `context.indexSettings().hasValue(INDEX_KNN_SPACE_TYPE.getKey())` returns false with defaults. [Here](https://github.com/jmazanec15/k-NN-1/blob/mapping-refactor/src/test/java/com/amazon/opendistroforelasticsearch/knn/index/KNNESIT.java#L147) is test case